storage.go

```diff
@@ -48,10 +48,30 @@ type Storage interface {
 
 	// InitialState returns the saved HardState and ConfState information.
 	InitialState() (pb.HardState, pb.ConfState, error)
-	// Entries returns a slice of log entries in the range [lo,hi).
-	// MaxSize limits the total size of the log entries returned, but
-	// Entries returns at least one entry if any.
+
+	// Entries returns a slice of log entries in the range [lo, hi). The maxSize
+	// limits the total size of the log entries returned, but Entries returns at
+	// least one entry if any.
+	//
+	// Entries that storage returns via this method are not modified by the raft
+	// library, but can be forwarded back to the application via Ready struct. The
+	// ownership of these entries is thus decided between Storage implementation
+	// and the application. It is recommended that the application does not modify
+	// them either way, but if it needs to do so, it must be aware that this can
+	// affect the internal state of the Storage.
+	//
+	// The raft library or application can append to the returned slice though.
+	// The Storage implementation must protect its state from corruption that such
+	// appends may cause. Two common ways to do so:
+	//  - allocate the slice before returning it, and assume that slice ownership
+	//  is transferred to this library and the application (the safest option),
+	//  - use Go full slice expressions before returning the slice, so that
+	//  appends will reallocate/copy it (see MemoryStorage).
+	//
+	// Returns ErrCompacted if entry lo has been compacted, or ErrUnavailable if
```

Comment:
How accurate is this? I'm seeing that in `MemoryStorage`, `ErrUnavailable` is only returned if we haven't got any entries after the snapshot. But we also verify that `hi <= ms.lastIndex()+1`, and so is `lo`, so technically it seems that we return `ErrUnavailable` only if the requested `lo == hi`. This shouldn't be the only case though.

We should clarify these semantics, and fix `MemoryStorage` to be a reference implementation.

Reply:
https://github.com/cockroachdb/cockroach/blob/8f975693228f7a676204594540973de9494c54fb/pkg/kv/kvserver/logstore/logstore.go#L576-L585

https://github.com/cockroachdb/cockroach/blob/8f975693228f7a676204594540973de9494c54fb/pkg/kv/kvserver/logstore/logstore.go#L600-L601

So in CRDB, we get `ErrUnavailable` only if we're trying to scan `[lo,hi)` and we're finding that not all of the entries we are trying to return are there. For example, if the log is 5,6,7 we would return this error from `[5,8)`, `[1,4)`, `[1,6)`.

The `MemoryStorage` will instead panic if `hi >= lastIndex` and so returns `ErrUnavailable` only if we start to scan from below the first log index.

For a reference impl it would probably be better to match the CRDB behavior, i.e. return `ErrUnavailable` either way - this _should_ be a dead code path (because why would raft read past the last index?) but it's easier if the implementer can just handle these cases the same way.